File: sol/solution/0001-0100/0039. Combination Sum/Solution.ex
defmodule Solution do
  @spec combination_sum(candidates :: [integer], target :: integer) :: [[integer]]
  def combination_sum(candidates, target) do
    candidates
    |> Enum.sort()
    |> backtrack(target, [], 0)
  end
  defp backtrack(candidates, 0, current, _index) do
    [Enum.reverse(current)]
  end
  defp backtrack(candidates, target, current, index) when target < 0 do
    []
  end
  defp backtrack(candidates, target, current, index) do
    Enum.flat_map(index..(length(candidates) - 1), fn i ->
      c = Enum.at(candidates, i)
      backtrack(candidates, target - c, [c | current], i)
    end)
  end
end

File: sol/solution/0001-0100/0039. Combination Sum/Solution.erl
-spec combination_sum(Candidates :: [integer()], Target :: integer()) -> [[integer()]].
combination_sum(Candidates, Target) ->
    Sorted = lists:sort(Candidates),
    combinations(Sorted, Target, []).
combinations(_Candidates, 0, Current) ->
    [lists:reverse(Current)];
combinations([], _Target, _Current) ->
    [];
combinations([C | Rest] = Candidates, Target, Current) ->
    if
        C > Target ->
            [];
        true ->
            WithC = combinations(Candidates, Target - C, [C | Current]),
            WithoutC = combinations(Rest, Target, Current),
            WithC ++ WithoutC
    end.